#36 Update path to csw endpoint (#37)
metno/data-management-handbook: doc/search_context.adoc

===== on the online catalog

For searching the online metadata catalog, please refer to `https://csw.s-enda.k8s.met.no/` instead of `http://10.20.30.10/`.

===== OpenSearch examples

To find all datasets in the catalog:

* `http://10.20.30.10/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results`

Or datasets within a given time span:

* `http://10.20.30.10/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results&time=2000-01-01/2020-09-01`

[note]
====
pyCSW opensearch time queries do not relate to the "expected" `start_date` and `end_date` MMD fields, or the associated `beginPosition` and `endPosition`, in the pyCSW ISO19139 profile. We are working to understand and handle this...

> ====

Author: Yes, I just tested it, the dates work as expected now. I'll remove the note.

===== Online catalog

For searching the online metadata catalog, the base url (`http://10.20.30.10/`) must be replaced by `https://csw.s-enda.k8s.met.no/`.

===== OpenSearch examples

To find all datasets in the catalog:

* `https://csw.s-enda.k8s.met.no/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results`

Or datasets within a given time span:

* `http://csw.s-enda.k8s.met.no/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results&time=2000-01-01/2020-09-01`

Or datasets within a geographical domain (defined as a box with parameters min_longitude, min_latitude, max_longitude, max_latitude):

* `https://csw.s-enda.k8s.met.no/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results&bbox=0,40,10,60`

Or, datasets from any of the Sentinel satellites:

* `https://csw.s-enda.k8s.met.no/?mode=opensearch&service=CSW&version=2.0.2&request=GetRecords&elementsetname=full&typenames=csw:Record&resulttype=results&q=sentinel`

==== Advanced geographical search

> Pycsw only supports geographical searches querying for a box. For more advanced geographical searches, one must write his own XML files. For example:

Reviewer: ```suggestion
PyCSW opensearch only supports geographical searches querying for a box. For more advanced geographical searches, one must write specific XML files. For example:
```

> * to find all datasets that contains a point

Reviewer: ```suggestion
* To find all datasets containing a point:
```

[source, xml]
----
<?xml version="1.0" encoding="ISO-8859-1" standalone="no"?>
<csw:GetRecords
        xmlns:csw="http://www.opengis.net/cat/csw/2.0.2"
        xmlns:ogc="http://www.opengis.net/ogc"
        xmlns:gml="http://www.opengis.net/gml"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        service="CSW" version="2.0.2" resultType="results" maxRecords="10" outputFormat="application/xml" outputSchema="http://www.opengis.net/cat/csw/2.0.2" xsi:schemaLocation="http://www.opengis.net/cat/csw/2.0.2 http://schemas.opengis.net/csw/2.0.2/CSW-discovery.xsd" >
    <csw:Query typeNames="csw:Record">
        <csw:ElementSetName>full</csw:ElementSetName>
        <csw:Constraint version="1.1.0">
            <ogc:Filter>
                <ogc:Contains>
		            <ogc:PropertyName>ows:BoundingBox</ogc:PropertyName>
			        <gml:Point>
                	        <gml:pos srsDimension="2">59.0 4.0</gml:pos>
                    </gml:Point>
                </ogc:Contains>
            </ogc:Filter>
        </csw:Constraint>
    </csw:Query>
</csw:GetRecords>
----

> * to find all datasets that intersects a polygon

Reviewer: ```suggestion
* To find all datasets intersecting a polygon:
```

        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        service="CSW" version="2.0.2" resultType="results" maxRecords="10" outputFormat="application/xml" outputSchema="http://www.opengis.net/cat/csw/2.0.2" xsi:schemaLocation="http://www.opengis.net/cat/csw/2.0.2 http://schemas.opengis.net/csw/2.0.2/CSW-discovery.xsd" >
    <csw:Query typeNames="csw:Record">
        <csw:ElementSetName>full</csw:ElementSetName>
        <csw:Constraint version="1.1.0">
            <ogc:Filter>
                <ogc:Intersects>
		            <ogc:PropertyName>ows:BoundingBox</ogc:PropertyName>
                      <gml:Polygon>
                           <gml:exterior>
                                <gml:LinearRing>
                                    <gml:posList>
                                        47.00 -5.00 55.00 -5.00 55.00 20.00 47.00 20.00 47.00 -5.00
                                    </gml:posList>
                                </gml:LinearRing>
                           </gml:exterior>
                    </gml:Polygon>
                </ogc:Intersects>
            </ogc:Filter>
        </csw:Constraint>
    </csw:Query>
</csw:GetRecords>
----

> * How to use the XML file to query the CSW endpoint with python:

Reviewer: ```suggestion
* Then, you can query the CSW endpoint with, e.g., python:
```

[source, python]
----
import requests
requests.post('https://csw.s-enda.k8s.met.no', data=open(my_xml_request).read()).text
----

==== Web portals

*GeoNorge.no*

*TODO:* describe how to search in geonorge, possibly with screenshots

==== QGIS

> The production server for the CSW catalog service is available at `https://csw.s-enda.k8s.met.no`. This can be used from QGIS as follows:

Reviewer: ```suggestion
MET Norway's S-ENDA CSW catalog service is available at `https://csw.s-enda.k8s.met.no`. This can be used from QGIS as follows:
```

[[search_context]]
=== Search context

==== Using OpenSearch

===== Local test machines

The `localtest` and `localdev` virtual machines provide OpenSearch support through link:https://github.com/geopython/pycsw[PyCSW]. To test OpenSearch via the browser, start the `localtest` vm (`vagrant up localtest`) and go to the following address:

* `http://10.20.30.10/pycsw/csw.py?mode=opensearch&service=CSW&version=2.0.2&request=GetCapabilities`

> This will return a description document of the catalog service. The `URL` field in the description document is a template format that can be used to represent a parameterized form of the search. The search client will process the URL template and attempt to replace each instance of a template parameter, generally represented in the form {name}, with a value determined at query time (link:https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#opensearch-url-template-syntax[OpenSearch URL template syntax]). The question mark following any search parameter means that the parameter is optional.

Reviewer: This part should be updated when @aheimsbakk is done with the new vagrant. Maybe today. It could be done in this PR, otherwise let's make a new issue :)

Reviewer: we'll do this in another issue

> ==== Online catalog

Reviewer: I think this could be subsection of section "6.3.1 User searches the catalog with OpenSearch". Can that section also be renamed to "OpenSearch"? The point should be demonstrate the most common search queries. As such, we should perhaps also add an example of a geographical search, as the users are probably most interested in searching in time and space.

The virtual machines are now being changed as well but we can update the text for that when we're done with (or as part of) https://app.zenhub.com/workspaces/s-enda-5e7b1b4acfc5ca185fb6b53c/issues/metno/s-enda-project-plan/428

Author: I have re-organized a bit the OpenSearch section to try and make it more clear. What do you think?

==== User searches from QGIS

> A staging server for the CSW catalog service is available at link:https://csw-staging.s-enda.k8s.met.no. This can be used from QGIS, as follows:

Reviewer: let's use the production server in this example

Author: I agree that in the end it should be the production server. But is it already up? I used this `https://gitlab.met.no/tjenester/s-enda` as a reference and it says that the production endpoint is not available yet.

Reviewer: No, it's not up yet. I was thinking that we could just write this as if it is up, though. Then we don't need to change it later :)

. Select `Web > MetaSearch > MetaSearch` menu item
. Select `Services > New`
. Type, e.g., `csw-staging-senda` for the name

> . Type `https://csw-staging.s-enda.k8s.met.no` for the URL

Reviewer: production server
